src/Adapter/Order/CommandHandler/UpdateProductInOrderHandler.php

```diff
@@ -186,4 +217,164 @@ private function assertProductCanBeUpdated(
             }
         }
     }
+
+    /**
+     * @param Number $priceTaxIncluded
+     * @param Number $priceTaxExcluded
+     * @param Order $order
+     * @param Cart $cart
+     * @param Product $product
+     * @param Combination|null $combination
+     */
+    private function createSpecificPriceIfNeeded(
+        Number $priceTaxIncluded,
+        Number $priceTaxExcluded,
+        Order $order,
+        Cart $cart,
+        Product $product,
+        $combination
+    ): void {
+        // Check it the SpecificPrice has already been added by restoreOrderProductsSpecificPrices, if yes ignore new
+        // price because the first one is kept
+        if (SpecificPrice::exists(
+            $product->id,
+            $combination ? $combination->id : 0,
+            0,
+            0,
+            0,
+            $order->id_currency,
+            $order->id_customer,
+            1,
+            DateTime::NULL_VALUE,
+            DateTime::NULL_VALUE
+        )) {
+            return;
+        }
+
+        $initialProductPriceTaxExcl = Product::getPriceStatic(
+            $product->id,
+            false,
+            $combination ? $combination->id : null,
+            $this->getPrecisionFromCart($cart),
+            null,
+            false,
+            true,
+            1,
+            false,
+            $order->id_customer,
+            $cart->id,
+            $order->{Configuration::get('PS_TAX_ADDRESS_TYPE', null, null, $order->id_shop)}
+        );
+
+        // Better check with price tax excluded since it's the one saved in database, if the price matches
+        // the product's one no need for specific price
+        if ($priceTaxExcluded->equals(new Number((string) $initialProductPriceTaxExcl))) {
+            return;
+        }
+
+        $specificPrice = new SpecificPrice();
+        $specificPrice->id_shop = 0;
+        $specificPrice->id_cart = 0;
+        $specificPrice->id_shop_group = 0;
+        $specificPrice->id_currency = $order->id_currency;
+        $specificPrice->id_country = 0;
+        $specificPrice->id_group = 0;
+        $specificPrice->id_customer = $order->id_customer;
+        $specificPrice->id_product = $product->id;
+        $specificPrice->id_product_attribute = $combination ? $combination->id : 0;
+        $specificPrice->price = (float) (string) $priceTaxExcluded;
+        $specificPrice->from_quantity = 1;
+        $specificPrice->reduction = 0;
+        $specificPrice->reduction_type = 'amount';
+        $specificPrice->reduction_tax = !$priceTaxIncluded->equals($priceTaxExcluded);
+        $specificPrice->from = '0000-00-00 00:00:00';
+        $specificPrice->to = '0000-00-00 00:00:00';
+        $specificPrice->add();
+        $this->temporarySpecificPrices[] = $specificPrice;
+    }
+
+    /**
+     * @param Order $order
+     *
+     * @return Cart
+     */
+    private function createNewOrEditExistingCart(Order $order)
```

Comment:
I'm really not confident about this "create a cart maybe.." thing I know we hav the same in AddPorductToOrder but it's still not normal to create a cart Franck manage to get rid of this cart duplication so we should not allow this

An order MUST have a cart since it's created based on a cart anyway so I don't understand in which case there could be no cart to retrieve..